### PR TITLE
[DependencyInjection] Fix serialization of \Closure in RemoveUnusedDefinitionsPass

### LINKED ISSUE #29694: [DependencyInjection] Regression? - Serialization of 'Closure' is not allowed
**Symfony version(s) affected**: 4.?

**Description**  
Have just updated symfony and now getting an error "Fatal error: Uncaught Exception: Serialization of 'Closure' is not allowed in symfony\dependency-injection\Compiler\RemoveUnusedDefinitionsPass.php"

I *believe* I may be hitting the issue due to the the following change:

https://github.com/symfony/dependency-injection/commit/465af9012190665606f1355c34f508b12faf1ad4#diff-4995ff3c23fc152e94bb1d845e198c62L73

However, the cause is as follows:

1. RemoveUnusedDefintions.php calls 'serialize' on a Definition object
2. Definition object stores a list of 'errors'. This is set in lines 885-888 of Definition.php in the addError function; which takes a string|\Closure
3. Trying to serialize a dependency with a previous error generates due to the \Closure argument

### PR DESCRIPTION
**Pull request by XuruDragon**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no 
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #29694 
| License       | MIT
| Doc PR        | n/a

Fix the issue #29694

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
If I remember correctly, we're using closures here to run their logic lazily, because doing it earlier has other unwanted side effects.
If that's confirmed (GitHub history to the rescue) then this patch is not correct.
Could you please have a look?

**Comment by XuruDragon** [comment]
ok, will have a look

**Comment by nicolas-grekas** [comment]
That could work! A test case would definitely prove it :)

File: src/Symfony/Component/DependencyInjection/Definition.php [hunk 1]
```diff
@@ -68,6 +68,41 @@ public function __construct($class = null, array $arguments = [])
...
     }
 
+    public function __sleep()
+    {
+        if ($this->hasErrors()) {
+            //By calling it, we resolve all potentials \Closure inside $this->>errors
+            $this->getErrors();
+        }
+
+        return [
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Definition.php]: array_keys(get_object_vars())?

**Comment by XuruDragon** [comment]
I've to change some, __sleep have a strange behavior look at appveyor result...

So I've change to the \Serializable interface and I'm currently implementing the two methods

File: src/Symfony/Component/DependencyInjection/Definition.php (hunk 1 above)
Comment (XuruDragon) [reply src/Symfony/Component/DependencyInjection/Definition.php]: yes I've tested that but look à my comment under

**Comment by nicolas-grekas** [comment]
We're trying to kill Serializable, we should try understand what's being this strangeness before resorting to the interface.

**Comment by nicolas-grekas** [comment]
Ok, it's not strange, it's the specifications of __sleep(), see its doc.
We need to use a dynamic property and implement __wakeup instead.

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @XuruDragon.

**Pull request merged**